File: src/nexus/models/agent.py
"""Agent model - the core autonomous employee entity."""

import uuid
from datetime import datetime, timezone
from typing import Any, Optional

from sqlmodel import Column, Field, Index, SQLModel
from sqlalchemy import JSON


class Agent(SQLModel, table=True):
    """An autonomous AI agent operating as an employee within the company.

    Agents have persistent identity, configurable runtime, skills, tools,
    budget constraints, and memory namespaces. They can be managed through
    lifecycle operations (hire, configure, wake, pause, terminate).
    """

    __tablename__ = "agents"
    __table_args__ = (
        Index("ix_agents_company_status", "company_id", "status"),
        Index("ix_agents_company_manager", "company_id", "manager_id"),
    )

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    company_id: uuid.UUID = Field(foreign_key="companies.id", index=True)
    name: str = Field(max_length=255)
    title: Optional[str] = Field(default=None, max_length=255)
    role: str = Field(max_length=100)
    department_id: Optional[uuid.UUID] = Field(
        default=None, foreign_key="departments.id"
    )
    team_id: Optional[uuid.UUID] = Field(default=None, foreign_key="teams.id")
    manager_id: Optional[uuid.UUID] = Field(
        default=None, foreign_key="agents.id"
    )

    # Status and lifecycle
    status: str = Field(default="idle", max_length=50)
    pause_reason: Optional[str] = Field(default=None)
    paused_at: Optional[datetime] = Field(default=None)
    error_reason: Optional[str] = Field(default=None)

    # Runtime configuration
    adapter_type: str = Field(default="langchain", max_length=100)
    adapter_config: Optional[dict[str, Any]] = Field(
        default=None, sa_column=Column(JSON)
    )
    runtime_config: Optional[dict[str, Any]] = Field(
        default=None, sa_column=Column(JSON)
    )
    model: Optional[str] = Field(default=None, max_length=255)

    # Capabilities and identity
    capabilities: Optional[list[str]] = Field(default=None, sa_column=Column(JSON))
    responsibilities: Optional[str] = Field(default=None)
    objectives: Optional[str] = Field(default=None)
    skills: Optional[list[str]] = Field(default=None, sa_column=Column(JSON))
    tools: Optional[list[str]] = Field(default=None, sa_column=Column(JSON))
    permissions: Optional[dict[str, Any]] = Field(
        default=None, sa_column=Column(JSON)
    )
    soul_description: Optional[str] = Field(default=None)

    # Budget and performance
    budget_monthly_cents: int = Field(default=0)
    spent_monthly_cents: int = Field(default=0)
    performance_score: Optional[float] = Field(default=None)

    # Memory
    memory_namespace: Optional[str] = Field(default=None, max_length=255)

    # Heartbeat and monitoring
    last_heartbeat_at: Optional[datetime] = Field(default=None)

    # Timestamps
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

File: dashboard/src/types/agent.ts
import type { AgentStatus, DateTimeString, UUID } from './common';

export interface Agent {
  id: UUID;
  company_id: UUID;
  name: string;
  title: string;
  role: string;
  department_id: UUID | null;
  team_id: UUID | null;
  manager_id: UUID | null;
  status: AgentStatus;
  adapter_type: string;
  model: string;
  capabilities: string[];
  responsibilities: string;
  objectives: string;
  budget_monthly_cents: number;
  spent_monthly_cents: number;
  performance_score?: number;
  soul_description: string;
  /** Per-action autonomy policy: { action_type: 1|2|3, spend_above_cents?: number }. */
  autonomy_policy?: Record<string, number> | null;
  last_heartbeat_at: DateTimeString | null;
  created_at: DateTimeString;
  updated_at: DateTimeString;
}

export interface AgentCreateRequest {
  name: string;
  title: string;
  role: string;
  department_id?: UUID;
  team_id?: UUID;
  manager_id?: UUID;
  adapter_type: string;
  model: string;
  capabilities?: string[];
  responsibilities?: string;
  objectives?: string;
  budget_monthly_cents?: number;
  soul_description?: string;
}

export interface AgentUpdateRequest {
  name?: string;
  title?: string;
  role?: string;
  department_id?: UUID | null;
  team_id?: UUID | null;
  manager_id?: UUID | null;
  status?: AgentStatus;
  adapter_type?: string;
  model?: string;
  capabilities?: string[];
  responsibilities?: string;
  objectives?: string;
  budget_monthly_cents?: number;
  soul_description?: string;
}

export interface MemoryEntry {
  id: UUID;
  company_id: UUID;
  agent_id: UUID | null;
  scope: string;
  scope_id: UUID | null;
  content: string;
  metadata: Record<string, unknown>;
  importance: number;
  access_count: number;
  tier: string;
  created_at: DateTimeString;
  updated_at: DateTimeString;
}
